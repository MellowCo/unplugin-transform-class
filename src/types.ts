import type { FilterPattern } from '@rollup/pluginutils'

export interface Options {
  /**
   * 自定义转换规则
   * @default
   * {
      '.': '-dl-',
      '/': '-sl-',
      ':': '-cl-',
      '%': '-pes-',
      '!': '-el-',
      '#': '-wn-',
      '(': '-lbl-',
      ')': '-lbr-',
      '[': '-lfl-',
      ']': '-lfr-',
      '$': '-do-',
      ',': '-lco-',
      '=': '-eqe-',
    }
   */
  rules?: Record<string, string>

  /**
   * 排除转换目标
   * @default [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/]
   */
  exclude?: FilterPattern

  /**
   * 需要转换的目标
   * @default [/\.[jt]sx?$/, /\.vue$/,  /\.vue\?vue/]
   */
  include?: FilterPattern
}
