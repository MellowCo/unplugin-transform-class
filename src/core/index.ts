/**
 * 获取class
 */
export function getClass(code: string) {
  const matchs: string[][] = []
  // vue
  Array.from(code.matchAll(/class="((?:\n|.)+?)"/g)).forEach((m) => {
    const classStr = m[1]
    let classArr = [m[0]]
    if (classStr.startsWith('{'))
      classArr = classArr.concat(getObjClass(classStr))
    else if (classStr.startsWith('['))
      classArr = classArr.concat(getArrClass(classStr))
    else
      classArr.push(classStr)
    matchs.push(classArr)
  })
  // react jsx tsx
  Array.from(code.matchAll(/className=["']((?:\n|.)+?)["']/g)).forEach((m) => {
    matchs.push([m[0], m[1]])
  })
  return matchs
}

export function getObjClass(className: string) {
  // class="{ 'bg-teal-200:55': title === 'Hello', 'h-1.000%': title === 'Hello' }"
  // => ['bg-teal-200:55','h-1.000%']
  return Array.from(className.matchAll(/'([^,]+?)'\s*:/g)).map(v => v[1])
}

export function getArrClass(className: string) {
  // [
  //   title === '2.3' ? 'font-$font-name bg-teal-200:55' :'tracking-[2/5]',
  //   isFont ? 'font-$font-name' : 'tracking-[2/5]'
  // ]
  // => ['font-$font-name bg-teal-200:55', 'tracking-[2/5]','font-$font-name tracking-[2/5]']
  return Array.from(className.matchAll(/(?<=[\?\:])\s*'(.+?)'/g)).map(v => v[1])
}

const transformRules: Record<string, string> = {
  '.': '-d-',
  '/': '-s-',
  ':': '-c-',
  '%': '-p-',
  '!': '-e-',
  '#': '-w-',
  '(': '-bl-',
  ')': '-br-',
  '[': '-fl-',
  ']': '-fr-',
  '$': '-r-',
}

const transformRegExp = /[\.\/:%!#\(\)\[\]$]/

export function transformCode(code: string) {
  const classNames = getClass(code)
  classNames.forEach((c) => {
    let currentClass = c[0]
    c.slice(1).forEach((selector) => {
      currentClass = currentClass.replace(selector, transformSelector(selector))
    })
    code = code.replace(c[0], currentClass)
  })

  return code
}

export function transformSelector(selector: string) {
  const escapePrefix = '\\'

  if (transformRegExp.test(selector)) {
    for (const transformRule in transformRules) {
      const replaceReg = new RegExp(`${escapePrefix}${transformRule}`, 'g')
      selector = selector.replace(replaceReg, transformRules[transformRule])
    }
  }

  return selector
}

