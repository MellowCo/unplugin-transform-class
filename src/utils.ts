export const defaultRules: Record<string, string> = {
  '.': '_dl_',
  '/': '_sl_',
  ':': '_cl_',
  '%': '_pes_',
  '!': '_el_',
  '#': '_wn_',
  '(': '_lbl_',
  ')': '_lbr_',
  '[': '_lfl_',
  ']': '_lfr_',
  '$': '_do_',
  ',': '_lco_',
  '=': '_eqe_',
  '+': '_plus_',
  '*': '_star_',
}

export function escapeRegExp(str = '') {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

function createTransformRegExp(rules: Record<string, string>, needEscape = false) {
  const escapePrefix = '\\'
  return new RegExp(Object.keys(rules).map(rule => escapeRegExp(`${needEscape ? escapePrefix : ''}${rule}`)).join('|'), 'g')
}

/**
 * 转换选择器字符
 * @param selector 选择器
 * @param rules 转换规则
 * @example bg-[#452233]:50 => bg--fl--w-452233-fr--c-40-c-50
 */
export function transformSelector(selector = '', rules = defaultRules) {
  const transformRegExp = createTransformRegExp(rules)

  return selector.replaceAll(transformRegExp, (m) => {
    return rules[m]
  })
}

export const cacheTransformSelector = (function () {
  let transformRegExp: RegExp

  return (selector = '', rules = defaultRules) => {
    if (!transformRegExp)
      transformRegExp = createTransformRegExp(rules)

    return selector.replaceAll(transformRegExp, (m) => {
      return rules[m]
    })
  }
}())

/**
 * 转换转义的选择器字符
 * @param selector 转义选择器
 * @param rules 转换规则
 * @example bg-\[\#452233\]\:50 => bg--fl--w-452233-fr--c-40-c-50
 */
export function transformEscapESelector(selector = '', rules = defaultRules) {
  const transformRegExp = createTransformRegExp(rules, true)

  return selector.replaceAll(transformRegExp, (m) => {
    return rules[m.replace('\\', '')]
  })
}

export const cacheTransformEscapESelector = (function () {
  let transformRegExp: RegExp

  return (selector = '', rules = defaultRules) => {
    if (!transformRegExp)
      transformRegExp = createTransformRegExp(rules, true)

    return selector.replaceAll(transformRegExp, (m) => {
      return rules[m.replace('\\', '')]
    })
  }
}())

/**
 * 还原转换后的选择器
 * @param selector 选择器
 * @param rules 转换规则
 * @example .bg--fl--w-452233-fr--c-40-c-50 => .bg-[#452233]:50
 */
export function restoreSelector(selector = '', rules = defaultRules) {
  const reverseRules = Object.fromEntries(Object.entries(rules).map(([key, value]) => [value, key]))
  const transformRegExp = createTransformRegExp(reverseRules)

  return selector.replaceAll(transformRegExp, (m) => {
    return reverseRules[m]
  })
}

export const cacheRestoreSelector = (function () {
  let transformRegExp: RegExp
  let reverseRules: Record<string, string>

  return (selector = '', rules = defaultRules) => {
    if (!transformRegExp) {
      reverseRules = Object.fromEntries(Object.entries(rules).map(([key, value]) => [value, key]))
      transformRegExp = createTransformRegExp(reverseRules)
    }

    return selector.replaceAll(transformRegExp, (m) => {
      return reverseRules[m]
    })
  }
}())

export { transformCode, getClass } from './core'
