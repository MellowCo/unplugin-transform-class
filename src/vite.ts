import unplugin from '.'

export { transformSelector } from './core'

export const transformWxClass = unplugin.vite
