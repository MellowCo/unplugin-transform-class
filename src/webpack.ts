import type { Options } from './types'
import unplugin from '.'

// TODO: some upstream lib failed generate invalid dts, remove the any in the future
export default unplugin.webpack as (options?: Options) => any
