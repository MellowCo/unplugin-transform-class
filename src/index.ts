import { createUnplugin } from 'unplugin'
import type { Options } from './types'
import { transformCode } from './core'

export default createUnplugin<Options>(() => ({
  name: 'unplugin-transform-we-class',
  enforce: 'pre',
  transformInclude(id) {
    return /.(vue|jsx|tsx)$/.test(id)
  },
  transform(code) {
    return transformCode(code)
  },
}))
