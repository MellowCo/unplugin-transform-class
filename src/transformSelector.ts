import {transformSelector} from './core'

export default transformSelector
