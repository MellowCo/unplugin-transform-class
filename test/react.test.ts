/*
 * @Author: licl
 * @Date: 2022-07-08 15:56:57
 * @LastEditTime: 2022-07-23 09:08:28
 * @LastEditors: licl
 * @Description:
 */
import { describe, expect, it } from 'vitest'
import { getClass, transformCode } from '../src/core'
import { dynamicCode, emptyClassName, reactCode, xxClassName } from './assets/react'

describe('react', () => {
  it('getClass', () => {
    expect(getClass(reactCode)).toMatchSnapshot()
    expect(getClass(dynamicCode)).toMatchSnapshot()
    expect(getClass(emptyClassName)).toMatchSnapshot()
    expect(getClass(xxClassName)).toMatchSnapshot()
  })

  it('transformCode', () => {
    expect(transformCode(reactCode)).toMatchSnapshot()
    expect(transformCode(dynamicCode)).toMatchSnapshot()
    expect(transformCode(emptyClassName)).toMatchSnapshot()
    expect(transformCode(xxClassName)).toMatchSnapshot()
  })

})
