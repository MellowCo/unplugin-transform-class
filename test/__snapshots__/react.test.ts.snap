// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`react > getClass 1`] = `
[
  [
    "className='content'",
    "content",
  ],
  [
    "className='indent-1/2'",
    "indent-1/2",
  ],
  [
    "className='font-[system-ui] bg-teal-200:55'",
    "font-[system-ui] bg-teal-200:55",
  ],
  [
    "className='font-$font-name'",
    "font-$font-name",
  ],
  [
    "className='tracking-[2/5] bg-teal-200:55'",
    "tracking-[2/5] bg-teal-200:55",
  ],
  [
    "className='h-1.000%'",
    "h-1.000%",
  ],
  [
    "className='bg-teal-200:55'",
    "bg-teal-200:55",
  ],
  [
    "className='text-area'",
    "text-area",
  ],
  [
    "className='text-gray-500_50'",
    "text-gray-500_50",
  ],
  [
    "className='text-gray-500:50'",
    "text-gray-500:50",
  ],
  [
    "className='text-gray-500/50'",
    "text-gray-500/50",
  ],
  [
    "className='text-red-500_50'",
    "text-red-500_50",
  ],
  [
    "className='text-red-500:50'",
    "text-red-500:50",
  ],
  [
    "className='text-red-500/50'",
    "text-red-500/50",
  ],
  [
    "className='text-yellow-500_50'",
    "text-yellow-500_50",
  ],
  [
    "className='text-yellow-500:50'",
    "text-yellow-500:50",
  ],
  [
    "className='text-yellow-500/50'",
    "text-yellow-500/50",
  ],
  [
    "className='text-blue-500_50'",
    "text-blue-500_50",
  ],
  [
    "className='text-blue-500/50'",
    "text-blue-500/50",
  ],
  [
    "className='text-blue-500:50'",
    "text-blue-500:50",
  ],
  [
    "className='text-green-500_50'",
    "text-green-500_50",
  ],
  [
    "className='text-green-500:50'",
    "text-green-500:50",
  ],
  [
    "className='text-green-500/50'",
    "text-green-500/50",
  ],
  [
    "className='c-hex-157_60'",
    "c-hex-157_60",
  ],
  [
    "className='c-hex-157/60'",
    "c-hex-157/60",
  ],
  [
    "className='c-[#157]/60'",
    "c-[#157]/60",
  ],
  [
    "className='c-[#157]_60'",
    "c-[#157]_60",
  ],
  [
    "className='c-[#157]:60'",
    "c-[#157]:60",
  ],
]
`;

exports[`react > transformCode 1`] = `
"<View className='content'>
<View className='indent-1_sl_2'> indent-1/2 </View>
<View className='font-_lfl_system-ui_lfr_ bg-teal-200_cl_55'> font-[system-ui] </View>
<View className='font-_do_font-name'> font-$font-name </View>

<View className='tracking-_lfl_2_sl_5_lfr_ bg-teal-200_cl_55'> tracking-[2/5] </View>
<View className='h-1_dl_000_pes_'> h-1.000% </View>

<View className='bg-teal-200_cl_55'> bg-teal-200:55 </View>
<View className='text-area'>

</View>

<View className='text-gray-500_50'>
text-gray-500
</View>

<View className='text-gray-500_cl_50'>
text-gray-500
</View>

<View className='text-gray-500_sl_50'>
text-gray-500
</View>

<View className='text-red-500_50'>
text-red-500
</View>

<View className='text-red-500_cl_50'>
text-red-500
</View>

<View className='text-red-500_sl_50'>
text-red-500
</View>

<View className='text-yellow-500_50'>
text-yellow-500
</View>

<View className='text-yellow-500_cl_50'>
text-yellow-500
</View>

<View className='text-yellow-500_sl_50'>
text-yellow-500
</View>

<View className='text-blue-500_50'>
text-blue-500
</View>

<View className='text-blue-500_sl_50'>
text-blue-500
</View>

<View className='text-blue-500_cl_50'>
text-blue-500
</View>

<View className='text-green-500_50'>
text-green-500
</View>

<View className='text-green-500_cl_50'>
text-green-500
</View>

<View className='text-green-500_sl_50'>
text-green-500
</View>

<View className='c-hex-157_60'>
c-hex-157
</View>

<View className='c-hex-157_sl_60'>
c-hex-157
</View>

<View className='c-_lfl__wn_157_lfr__sl_60'>
c-[#157]
</View>

<View className='c-_lfl__wn_157_lfr__60'>
c-[#157]
</View>

<View className='c-_lfl__wn_157_lfr__cl_60'>
c-[#157]
</View>
</View>"
`;
